client/chat/chat-reducer.ts

```diff
@@ -64,248 +66,271 @@ function updateMessages(
 
-      return users.set('active', active).set('idle', idle).set('offline', offline)
-    })
+    const channel = state.byName.get(lowerCaseChannelName)
+    if (!channel) {
+      return
+    }
+
+    channel.users.idle.set(user.id, user)
+    channel.users.active.delete(user.id)
+    channel.users.offline.delete(user.id)
   },
 
   ['@chat/updateUserOffline'](state, action) {
-    const { channel, user } = action.payload
-    const lowerCaseChannel = channel.toLowerCase()
-    return state.updateIn(['byName', lowerCaseChannel, 'users'], usersUncasted => {
-      // TODO(tec27): Remove cast once Immutable infers types properly
-      const users = usersUncasted as Users
-      const [offline, active, idle] = updateUserState(user, users.offline, users.active, users.idle)
-
-      return users.set('active', active).set('idle', idle).set('offline', offline)
-    })
+    const { channel: channelName, user } = action.payload
+    const lowerCaseChannelName = channelName.toLowerCase()
+
+    const channel = state.byName.get(lowerCaseChannelName)
+    if (!channel) {
+      return
+    }
+
+    channel.users.offline.set(user.id, user)
+    channel.users.active.delete(user.id)
+    channel.users.idle.delete(user.id)
   },
 
   ['@chat/loadMessageHistoryBegin'](state, action) {
-    const { channel } = action.payload
-    return state.setIn(['byName', channel.toLowerCase(), 'loadingHistory'], true)
+    const { channel: channelName } = action.payload
+    const lowerCaseChannelName = channelName.toLowerCase()
+
+    const channel = state.byName.get(lowerCaseChannelName)
+    if (!channel) {
+      return
+    }
+
+    channel.loadingHistory = true
   },
 
   ['@chat/loadMessageHistory'](state, action) {
     if (action.error) {
       // TODO(2Pac): Handle errors
-      return state
+      return
     }
 
-    const { channel, limit } = action.meta
-    const lowerCaseChannel = channel.toLowerCase()
-    const newMessages = List(
-      action.payload.map(
-        msg =>
-          new TextMessageRecord({
-            id: msg.id,
-            time: msg.sent,
-            from: msg.user.id,
-            text: msg.data.text,
-          }),
-      ),
-    )
-    let updated = state.setIn(['byName', lowerCaseChannel, 'loadingHistory'], false)
-    if (newMessages.size < limit) {
-      updated = updated.setIn(['byName', lowerCaseChannel, 'hasHistory'], false)
+    const { channel: channelName, limit } = action.meta
+    const lowerCaseChannelName = channelName.toLowerCase()
+
+    const channel = state.byName.get(lowerCaseChannelName)
+    if (!channel) {
+      return
     }
 
-    return updateMessages(updated, channel, false, messages => newMessages.concat(messages))
+    // Even though the payload here is `ServerChatMessage`, we expand its type so it can be
+    // concatenated with the existing messages which could also contain client chat messages.
+    const newMessages = action.payload as ChatMessage[]
```

Comment:
kinda surprised this would be necessary to do at all, is it because the type field is a different enum or something?

Reply:
I don't know the exact reason to be perfectly honest. Your theory sounds likely to me though.